### PR TITLE
Canopy tutorials

### PR DESCRIPTION
**Pull request by Espeer5**
## Purpose 
 
Add canopy standalone and coupled hydraulics tutorials to ClimaLSM/docs. See issues #240, #255

## To-do

- Add links to dynamics docs at locations marked with TODO once dynamic docs are live.

## Content

- Two usage tutorials included in docs/tutorials/Canopy

----
- [x] I have read and checked the items on the review checklist.

### CODE REVIEW

**Review by AlexisRenchon — APPROVED** [review]
Hi @Espeer5, your tutorial looks great, thanks!! Of course the obvious thing to fix is the dynamic docs link, but I imagine you are waiting for the actual merged into main docs link, which doesn't exist yet. 
I think what you can do for now is give the preview dynamic docs link, and add a comment on the dynamicdocs PR mentioning to correct those links when merged.

**Comment by kmdeck** [comment]
@Espeer5 I will review today! 

Just a note: before merging a pull request, after addressing review comments, we squash and rebase (to only merge a single commit with all the changes; it's a bit cleaner that way). I usually follow these directions: https://github.com/openedx/edx-platform/wiki/How-to-Rebase-a-Pull-Request.

File: docs/tutorials/Canopy/canopy_tutorial.jl
Comment (kmdeck) [inline docs/tutorials/Canopy/canopy_tutorial.jl:13]: couples -> combines or integrates

File: docs/tutorials/Canopy/canopy_tutorial.jl
Comment (kmdeck) [inline docs/tutorials/Canopy/canopy_tutorial.jl:14]: also a stomatal conductance model

File: docs/tutorials/Canopy/canopy_tutorial.jl
Comment (kmdeck) [inline docs/tutorials/Canopy/canopy_tutorial.jl:16]: "under either prescribed or simulated (as in a full Earth System Model) atmospheric and radiative flux conditions."

File: docs/tutorials/Canopy/canopy_tutorial.jl
Comment (kmdeck) [inline docs/tutorials/Canopy/canopy_tutorial.jl:22]: could we remove the TODO comments? We can add the sentence back in when the dynamic docs are live.

File: docs/tutorials/Canopy/canopy_tutorial.jl
Comment (kmdeck) [inline docs/tutorials/Canopy/canopy_tutorial.jl:29]: "The plant hydraulics model in ClimaLSM solves for the water content within bulk root-stem-canopy system using Richards equation discretized into an arbitrary number of layers. The water content is related to the water potential using a retention curve relationship, and the water potential is used to simulate the effect moisture stress has on transpiration and GPP."

File: docs/tutorials/Canopy/canopy_tutorial.jl
Comment (kmdeck) [inline docs/tutorials/Canopy/canopy_tutorial.jl:69]: and a stomatal conductance model

File: docs/tutorials/Canopy/canopy_tutorial.jl
Comment (kmdeck) [inline docs/tutorials/Canopy/canopy_tutorial.jl:80]: maybe explain that for 2D simulations, this would change, and can link to the ClimaLSM Domains documentation?

File: docs/tutorials/Canopy/canopy_tutorial.jl
Comment (kmdeck) [inline docs/tutorials/Canopy/canopy_tutorial.jl:183]: FT()

File: docs/tutorials/Canopy/canopy_tutorial.jl
Comment (kmdeck) [inline docs/tutorials/Canopy/canopy_tutorial.jl:210]: already defined above?

File: docs/tutorials/Canopy/canopy_tutorial.jl
Comment (kmdeck) [inline docs/tutorials/Canopy/canopy_tutorial.jl:282]: We are now using ClimaTimeSteppers - which does not support adaptive stepping.

File: docs/tutorials/Canopy/canopy_tutorial.jl
Comment (kmdeck) [inline docs/tutorials/Canopy/canopy_tutorial.jl:353]: "Model" to be consistent with GPP plot?

File: docs/tutorials/Canopy/soil_plant_hydrology_tutorial.jl
Comment (kmdeck) [inline docs/tutorials/Canopy/soil_plant_hydrology_tutorial.jl:3]: link to prev tutorial

File: docs/tutorials/Canopy/soil_plant_hydrology_tutorial.jl
Comment (kmdeck) [inline docs/tutorials/Canopy/soil_plant_hydrology_tutorial.jl:17]: "root extraction of the canopy hydraulics model, which acts as a boundary flux for the plant system, with a ['source term']..."

File: docs/tutorials/Canopy/soil_plant_hydrology_tutorial.jl
Comment (kmdeck) [inline docs/tutorials/Canopy/soil_plant_hydrology_tutorial.jl:74]: perhaps link to the Richards tutorial?

File: docs/tutorials/Canopy/soil_plant_hydrology_tutorial.jl
Comment (kmdeck) [inline docs/tutorials/Canopy/soil_plant_hydrology_tutorial.jl:102]: We might note that the SoilPlantHydrologyModel constructor will take these arguments and create the CanopyModel and RichardsModel instances internally.

File: docs/tutorials/Canopy/soil_plant_hydrology_tutorial.jl
Comment (kmdeck) [inline docs/tutorials/Canopy/soil_plant_hydrology_tutorial.jl:262]: Perhaps note that time-stepping Richards equation in the soil is done via an implicit method, while the canopy model can be stepped explicitly, so we use an IMEX (implicit-explicit) scheme for the combined model.

File: docs/tutorials/Canopy/soil_plant_hydrology_tutorial.jl
Comment (kmdeck) [inline docs/tutorials/Canopy/soil_plant_hydrology_tutorial.jl:291]: replace 0.0 with t0

Comment (kmdeck) [reply docs/tutorials/Canopy/soil_plant_hydrology_tutorial.jl:291]: (though you define t0 lower down)

**Review by kmdeck — COMMENTED** [review]
hi Edward, it looks good! I left a number of comments but the tutorials are in great shape.

File: docs/tutorials/Canopy/soil_plant_hydrology_tutorial.jl
Comment (kmdeck) [reply docs/tutorials/Canopy/soil_plant_hydrology_tutorial.jl:102]: (we can also note that this is done this way because the constructor has to enforce certain constraints on the models - e.g. that the root extraction is handled consistently between the two models - and in order to do this, the model creation must be done internally.)

**Comment by kmdeck** [comment]
> @Espeer5 I will review today!
> 
> Just a note: before merging a pull request, after addressing review comments, we squash and rebase (to only merge a single commit with all the changes; it's a bit cleaner that way). I usually follow these directions: https://github.com/openedx/edx-platform/wiki/How-to-Rebase-a-Pull-Request.

I can help you with this on Monday if you want - I usually make a backup branch before doing it just in case (rebasing changes the commit history, so there is the potential for lost code if something goes wrong).

**Comment by Espeer5** [comment]
> > @Espeer5 I will review today!
> > Just a note: before merging a pull request, after addressing review comments, we squash and rebase (to only merge a single commit with all the changes; it's a bit cleaner that way). I usually follow these directions: https://github.com/openedx/edx-platform/wiki/How-to-Rebase-a-Pull-Request.
> 
> I can help you with this on Monday if you want - I usually make a backup branch before doing it just in case (rebasing changes the commit history, so there is the potential for lost code if something goes wrong).

That's alright, I went ahead and did it this morning. I also have implemented the changes you mentioned in your comments.

**Review by kmdeck — APPROVED** [review]
Thanks for addressing the review comments!. I just noticed one change in soil_energy_hydrology tutorial which I think crept in accidentally. All set to merge then after reverting/squashing!

to merge, you comment "bors r+" on the conversation of the PR, which triggers the bors application. bors runs all the tests again and once they pass will merge the PR into main.

to cancel, for any reason, you can do "bors cancel" or "bors r-".

**Comment by kmdeck** [comment]
@Espeer5 not sure why buildkite failed in the bors run - it passed in github actions, and your changes didnt change the experiments folder at all. let's see if it passes this time..

**Pull request merged**
